internal/test_unit/python/unit.py

```diff
@@ -0,0 +1,371 @@
+                        cmd_outs += (f"\nERROR: ghostscript error for postscript file {filepath}")
+                        out_ok = False
+                
+                for filepath in test['out_exist']:
+                    # check output file exists and is nonzero size
+                    try:
+                        filesize = os.stat(filepath).st_size
+                        if filesize==0:
+                            cmd_outs += (f"\nERROR: file empty {filepath}\n")
+                            out_ok = False
+                            break
+                    except FileNotFoundError:
+                        cmd_outs += (f"\nERROR: file missing when it should exist {filepath}\n")
+                        out_ok = False
+
+                for filepath in test['out_not_exist']:
+                    # check output file doesn't exist
+                    if os.path.isfile(filepath):
+                        cmd_outs += (f"\nERROR: file exists when it should be missing {filepath}\n")
+                        out_ok = False
+
+            #   # unset the test environment variables
+            unset_envs = []
+            if 'env' in test.keys():
+                for key, val in sorted(test['env'].items()):
+                    del os.environ[key]
+                    unset_cmd = f"unset {key}"
+                    logger.debug(unset_cmd)
+                    unset_envs.append(unset_cmd)
+            
+            #   # print the test result
+            test_result = "pass" if (ret_ok and out_ok) else "FAIL"
+            logger.info(f"\t- {test_result} - \t{round(t_elaps.total_seconds(),3)} sec")
+
+            #   # on failure, print the problematic test and exit, if requested
+            if not (ret_ok and out_ok):
+                logger.info("\n".join(set_envs) + cmd + cmd_outs + "\n".join(unset_envs) + "\n")
+                if not noexit:
+                    sys.exit(1)
+
+
+def build_tests(test_root):
+    """
+    Parse the test components.
+
+    Take an ElementTree element extracted from a unit test xml file.
+    Return a list of all tests, where each test is represented as a dictionary,
+    with its keys representing each test component.
+
+    Parameters
+    ----------
+    test_root : ElementTree element
+        parsed from XML file containing the unit test(s) to perform
+
+    Returns
+    -------
+    test_list: 
+        list of test dicts, containing test attributes parsed from xml object
+
+    """
+
+    # define logger
+    logger = logging.getLogger(__name__)
+
+    # find all tests in test_xml, and create a dictionary of attributes for each test
+    test_list = []
+    for test_el in test_root.iter('test'):
+        test = {}
+        try:
+            test['name'] = test_el.attrib['name']
+        except KeyError:
+            logger.error("ERROR: name attribute not found for test")
+            raise
+            # should just fail this one test and not fully exit?
```

Comment:
In my opinion, it is OK to fully exit in this case, since it is a sign that something wasn't configured properly in the test definitions. It should be very obvious to a developer that they did something wrong instead of potentially missing the failure.